### PR TITLE
[14.0][IMP] shopfloor: location content transfer find_work multiple user

### PR DESCRIPTION
**Pull request by mt-software-de**
Before:
Also pickings were found which are already assigned to a user This leaded to two problems:
- if the qty done was already set to the max qty by an other user you always got the error no work
- or you got the same location and the same move lines as the user before

Now:
You only get pickings whitout a assigned user

### CODE REVIEW

**Comment by mt-software-de** [comment]
@jbaudoux this should solve the problem that the users getting always the same location until the previous picking isn't done

**Review by jbaudoux — APPROVED** [review]
Thanks for this fix.
To make it safe, there should also be a lock at the beginning of `_find_location_to_work_from` like here https://github.com/OCA/wms/blob/14.0/shopfloor_batch_automatic_creation/actions/picking_batch_auto_create.py#L59

**Comment by mt-software-de** [comment]
> Thanks for this fix. To make it safe, there should also be a lock at the beginning of `_find_location_to_work_from` like here https://github.com/OCA/wms/blob/14.0/shopfloor_batch_automatic_creation/actions/picking_batch_auto_create.py#L59

done

File: shopfloor/services/location_content_transfer.py [hunk 1]
```diff
@@ -245,18 +247,27 @@ def _create_moves_from_location(self, location):
...
-        )
-        move_lines = next_picking.move_line_ids.filtered(
-            lambda line: line.qty_done < line.product_uom_qty
-        )
-        return fields.first(move_lines).location_id
+        location = self.env["stock.location"]
+        offset = -1
+        while not location:
+            offset += 1
+            next_picking = self.env["stock.picking"].search(
```
Comment (simahawk) [inline shopfloor/services/location_content_transfer.py]: wouldn't be better to search them all just once and slice them w/ the offset?

File: shopfloor_base/utils.py [hunk 2]
```diff
@@ -65,3 +68,17 @@ def _get_running_env():
 
 
 RUNNING_ENV = _get_running_env()
+
+
+def _lock(cr, name):
```
Comment (simahawk) [inline shopfloor_base/utils.py]: I would move this to an action. We can have `self._actions_for("lock").advisory(name)`.

**Review by simahawk — CHANGES_REQUESTED** [review]

File: shopfloor_base/utils.py (hunk 2 above)
Comment (simahawk) [reply shopfloor_base/utils.py]: we can have `for_update` too and replace `_lock_lines` in cluster_picking and location_content_transfer

Comment (mt-software-de) [reply shopfloor_base/utils.py]: Thanks for the input i wasn't aware of  the actions, this a better place to put it in.
I created the new action and replaced the calls

File: shopfloor/services/location_content_transfer.py [hunk 3]
```diff
@@ -273,6 +280,7 @@ def find_work(self):
         response = self._recover_started_picking()
         if response:
             return response
+        self._actions_for("lock").advisory("find_work")
```
Comment (jbaudoux) [inline shopfloor/services/location_content_transfer.py]: Can the advisory lock name move to a variable or method? And I would give it a more unique name like location_content_transfer_find_work

File: shopfloor_base/actions/lock.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        """
+        hasher = hashlib.sha1(str(name).encode())
+        # pg_lock accepts an int8 so we build an hash composed with
+        # contextual information and we throw away some bits
+        int_lock = struct.unpack("q", hasher.digest()[:8])
+
+        self.env.cr.execute("SELECT pg_advisory_xact_lock(%s);", (int_lock,))
+        self.env.cr.fetchone()[0]
+
+    def for_update(self, table, ids, log_exceptions=False):
```
Comment (jbaudoux) [inline shopfloor_base/actions/lock.py]: I would have given the recordset in argument instead of table & ids

File: shopfloor_base/actions/lock.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2022 Michael Tietz (MT Software) <mtietz@mt-software.de>
```
Comment (simahawk) [inline shopfloor_base/actions/lock.py:1]:
```suggestion
# Copyright 2020 Camptocamp
# Copyright 2022 Michael Tietz (MT Software) <mtietz@mt-software.de>
```

**Review by simahawk — COMMENTED** [review]
LG really good, thanks.
One nitpicking thing: can you split the commit message pls?
Eg:
```
shopfloor_base: add actions for locking records

* advisory
* for_update
```

```
shopfloor: use lock actions
```

```
shopfloor: fix location content transfer find_work

Before:
Also pickings were found which are already assigned to a user
This leaded to two problems:
- if the qty done was already set to the max qty
  by an other user you always got the error no work
- or you got the same location and the same move lines as the user before

Now:
You only get pickings whithout a assigned user
```
The third one can contain the locking change for find_work, of course.

**Review by simahawk — APPROVED** [review]
Thanks for your improvements and your collaboration :+1:

**Pull request merged**